packages/core/src/types.ts

```diff
@@ -230,10 +230,13 @@ export type Receiver<TEvent extends EventObject> = (
   listener: (event: TEvent) => void
 ) => void;
 
-export type InvokeCallback<TEvent extends EventObject = AnyEventObject> = (
+export type InvokeCallback<
+  TEvent extends EventObject = AnyEventObject,
+  TReceivedEvent extends EventObject = AnyEventObject
+> = (
   callback: Sender<TEvent>,
-  onReceive: Receiver<TEvent>
-) => any;
+  onReceive: Receiver<TReceivedEvent>
+) => (() => void) | Promise<any> | void;
```

Comment:
Returning `Promise` can result in some very hard to debug issues - I actually spent some time recently figuring this out while helping out Kent.

So I wonder if maybe we shouldn't encourage this pattern in the types? I couldn't find a simple way to add a dev-only warning or anything about the issue that I was investigating.

Reply:
This is necessary because this pattern is supported:

```ts
src: () => async (sendBack, receive) => {
 // ...
}
```

The typings would fail if I remove `Promise<any>`, but I'm open to suggestions; I can't just omit it.

Reply:
This is exactly the problematic pattern that I've mentioned. Sorry for not being more specific earlier.

The problem here is that if the created promise resolves then:
- [this actor is removed from the parent](https://github.com/davidkpiano/xstate/blob/34ee27c008ca88798da1c4579fd94597f2467335/packages/core/src/interpreter.ts#L1010-L1018) and we no longer can't send any events to it (the `receive` won't be able to receive anything)
- `sendBack` should stop working (it probably doesn't stop working right now) because this actor has already sent `done.invoke.*` event to the parent and thus sending anything after that should be treated as invalid delivery

There is no easy way to distinguish `async` function from a regular one before actually calling it and inspecting its result (especially if you consider transpiled code as well). I think that in v5 we might provide both `fromCallback`, `fromAsync`, and `fromPromise` to differentiate between those patterns properly.

However, there is still a problem with the timing of the `async` completion - it feels natural to treat a resolved async actor as done but it effectively breaks the communication between it and the parent from that point in time. This sounds conceptually OK - if the actor has completed then no further communication should happen. It's a real gotcha though and I think that at least we should provide dev warnings for those cases but I'm unsure how this can be done effectively right now for `send(ev, { to: 'asyncActor' })`.

Reply:
> However, there is still a problem with the timing of the `async` completion - it feels natural to treat a resolved async actor as done but it effectively breaks the communication between it and the parent from that point in time. This sounds conceptually OK - if the actor has completed then no further communication should happen. It's a real gotcha though and I think that at least we should provide dev warnings for those cases but I'm unsure how this can be done effectively right now for `send(ev, { to: 'asyncActor' })`.

Let's discuss this in a separate issue/PR; seems tangential to this one.

For now, we have to keep `Promise<any>` in here just to make the types happy. We can refine in v5.

Reply:
@Andarist Should we make this a new issue, or is it blocking this PR?